File: agents/opencode-4-py/src/opencode_4_py/models/event.py
"""Event models."""

from __future__ import annotations

from pydantic import BaseModel, Field
from typing import Optional, List, Literal, Union, Dict, Any

from .session import Session, SessionStatus, Todo
from .message import Message, Part


class EventSessionCreated(BaseModel):
    """Session created event."""
    type: Literal["session.created"] = "session.created"
    properties: Dict[str, Any]


class EventSessionUpdated(BaseModel):
    """Session updated event."""
    type: Literal["session.updated"] = "session.updated"
    properties: Dict[str, Any]


class EventSessionDeleted(BaseModel):
    """Session deleted event."""
    type: Literal["session.deleted"] = "session.deleted"
    properties: Dict[str, Any]


class EventSessionStatus(BaseModel):
    """Session status event."""
    type: Literal["session.status"] = "session.status"
    properties: Dict[str, Any]


class EventSessionIdle(BaseModel):
    """Session idle event."""
    type: Literal["session.idle"] = "session.idle"
    properties: Dict[str, Any]


class EventSessionDiff(BaseModel):
    """Session diff event."""
    type: Literal["session.diff"] = "session.diff"
    properties: Dict[str, Any]


class EventSessionError(BaseModel):
    """Session error event."""
    type: Literal["session.error"] = "session.error"
    properties: Dict[str, Any]


class EventMessageUpdated(BaseModel):
    """Message updated event."""
    type: Literal["message.updated"] = "message.updated"
    properties: Dict[str, Any]


class EventMessageRemoved(BaseModel):
    """Message removed event."""
    type: Literal["message.removed"] = "message.removed"
    properties: Dict[str, Any]


class EventMessagePartUpdated(BaseModel):
    """Message part updated event."""
    type: Literal["message.part.updated"] = "message.part.updated"
    properties: Dict[str, Any]


class EventMessagePartDelta(BaseModel):
    """Message part delta event."""
    type: Literal["message.part.delta"] = "message.part.delta"
    properties: Dict[str, Any]


class EventPermissionAsked(BaseModel):
    """Permission asked event."""
    type: Literal["permission.asked"] = "permission.asked"
    properties: Dict[str, Any]


class EventPermissionUpdated(BaseModel):
    """Permission updated event."""
    type: Literal["permission.updated"] = "permission.updated"
    properties: Dict[str, Any]


class EventQuestionAsked(BaseModel):
    """Question asked event."""
    type: Literal["question.asked"] = "question.asked"
    properties: Dict[str, Any]


class EventTodoUpdated(BaseModel):
    """Todo updated event."""
    type: Literal["todo.updated"] = "todo.updated"
    properties: Dict[str, Any]


class EventFileEdited(BaseModel):
    """File edited event."""
    type: Literal["file.edited"] = "file.edited"
    properties: Dict[str, Any]


class EventFileWatcherUpdated(BaseModel):
    """File watcher updated event."""
    type: Literal["file.watcher.updated"] = "file.watcher.updated"
    properties: Dict[str, Any]


class EventServerConnected(BaseModel):
    """Server connected event."""
    type: Literal["server.connected"] = "server.connected"
    properties: Dict[str, Any]


class EventGlobalDisposed(BaseModel):
    """Global disposed event."""
    type: Literal["global.disposed"] = "global.disposed"
    properties: Dict[str, Any]


Event = Union[
    EventSessionCreated,
    EventSessionUpdated,
    EventSessionDeleted,
    EventSessionStatus,
    EventSessionIdle,
    EventSessionDiff,
    EventSessionError,
    EventMessageUpdated,
    EventMessageRemoved,
    EventMessagePartUpdated,
    EventMessagePartDelta,
    EventPermissionAsked,
    EventPermissionUpdated,
    EventQuestionAsked,
    EventTodoUpdated,
    EventFileEdited,
    EventFileWatcherUpdated,
    EventServerConnected,
    EventGlobalDisposed,
]


class GlobalEvent(BaseModel):
    """Global event."""
    directory: str
    payload: Event

File: sdk/oc4j/src/main/java/ai/opencode/sdk/model/event/Event.java
package ai.opencode.sdk.model.event;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

/**
 * Base class for all SSE events.
 * Uses Jackson polymorphic deserialization based on "type" field.
 * 
 * @see EventTypes for all supported event type IDs
 * @see EventType for enum representation
 */
@JsonTypeInfo(
    use = JsonTypeInfo.Id.NAME,
    include = JsonTypeInfo.As.PROPERTY,
    property = "type"
)
@JsonSubTypes({
    // Session events
    @JsonSubTypes.Type(value = EventSessionCreated.class, name = EventTypes.SESSION_CREATED),
    @JsonSubTypes.Type(value = EventSessionUpdated.class, name = EventTypes.SESSION_UPDATED),
    @JsonSubTypes.Type(value = EventSessionDeleted.class, name = EventTypes.SESSION_DELETED),
    @JsonSubTypes.Type(value = EventSessionStatus.class, name = EventTypes.SESSION_STATUS),
    @JsonSubTypes.Type(value = EventSessionIdle.class, name = EventTypes.SESSION_IDLE),
    @JsonSubTypes.Type(value = EventSessionDiff.class, name = EventTypes.SESSION_DIFF),
    @JsonSubTypes.Type(value = EventSessionError.class, name = EventTypes.SESSION_ERROR),
    // Message events
    @JsonSubTypes.Type(value = EventMessageUpdated.class, name = EventTypes.MESSAGE_UPDATED),
    @JsonSubTypes.Type(value = EventMessageRemoved.class, name = EventTypes.MESSAGE_REMOVED),
    @JsonSubTypes.Type(value = EventMessagePartUpdated.class, name = EventTypes.MESSAGE_PART_UPDATED),
    @JsonSubTypes.Type(value = EventMessagePartDelta.class, name = EventTypes.MESSAGE_PART_DELTA),
    // Permission events
    @JsonSubTypes.Type(value = EventPermissionAsked.class, name = EventTypes.PERMISSION_ASKED),
    @JsonSubTypes.Type(value = EventPermissionReplied.class, name = EventTypes.PERMISSION_REPLIED),
    // Question events
    @JsonSubTypes.Type(value = EventQuestionAsked.class, name = EventTypes.QUESTION_ASKED),
    // Todo events
    @JsonSubTypes.Type(value = EventTodoUpdated.class, name = EventTypes.TODO_UPDATED),
    // File events
    @JsonSubTypes.Type(value = EventFileEdited.class, name = EventTypes.FILE_EDITED),
    @JsonSubTypes.Type(value = EventFileWatcherUpdated.class, name = EventTypes.FILE_WATCHER_UPDATED),
    // Server events
    @JsonSubTypes.Type(value = EventServerConnected.class, name = EventTypes.SERVER_CONNECTED),
    @JsonSubTypes.Type(value = EventServerHeartbeat.class, name = EventTypes.SERVER_HEARTBEAT),
    // Global events
    @JsonSubTypes.Type(value = EventGlobalDisposed.class, name = EventTypes.GLOBAL_DISPOSED)
})
@JsonIgnoreProperties(ignoreUnknown = true)
public abstract class Event {
    private String type;

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }
}